paperoni/sql/collection.py

```diff
@@ -0,0 +1,608 @@
+        ],
+        "venue": [
+            ("paper", "paper_release", "paper_id"),
+            ("paper_release", "release", "release_id"),
+            ("release", "venue", "venue_id"),
+        ],
+        "release": [
+            ("paper", "paper_release", "paper_id"),
+            ("paper_release", "release", "release_id"),
+        ],
+    }
+
+    def __init__(self, filename):
+        self.db = Database(filename)
+        self.papers_added = set()
+        self.papers_excluded = set()
+
+    def __contains__(self, paper: Paper):
+        if paper in self.papers_excluded:
+            return False
+        elif paper in self.papers_added:
+            return True
+        else:
+            return self.db.select_id_from_values(
+                "paper",
+                "paper_id",
+                paper_id=self._find_paper_id(paper),
+                excluded=0,
+            )
+
+    def add(self, paper: Paper):
+        if paper in self.papers_excluded:
+            self.papers_excluded.remove(paper)
+        self.papers_added.add(paper)
+
+    def exclude(self, paper: Paper):
+        if paper in self.papers_added:
+            self.papers_added.remove(paper)
+        self.papers_excluded.add(paper)
+
+    def excludes(self, paper: Paper):
+        return paper in self.papers_excluded or self.db.select_id_from_values(
+            "paper",
+            "paper_id",
+            paper_id=self._find_paper_id(paper),
+            excluded=1,
+        )
+
+    def save(self):
+        try:
+            # Exclude papers.
+            for paper in self.papers_excluded:
+                self._create_paper_entry(paper, excluded=1)
+            # Add or update papers.
+            for paper in self.papers_added:
+                self._save_paper(paper)
+        except Exception as exc:
+            raise RuntimeError(
+                f"Error saving paper: {pprint.pformat(paper)}"
+            ) from exc
+
+    def _find_paper_id(self, paper: Paper):
+        # Check paper IDs then title+abstract.
+        for link_type in (
+            "SemanticScholar",
+            "MAG",
+            "ACL",
+            "DBLP",
+            "DOI",
+            "PubMed",
+            "PubMedCentral",
+        ):
+            paper_id = self.db.select_id_from_values(
+                "paper_link",
+                "paper_id",
+                link_type=link_type,
+                link=paper.get_ref(link_type),
+            )
+            if paper_id is not None:
+                return paper_id
+        return self.db.select_id_from_values(
+            "paper", "paper_id", title=paper.title, abstract=paper.abstract
+        )
+
+    def _find_author_id(self, author: Author):
+        # Check paper IDs then title+abstract.
+        for link_type in (
+            "SemanticScholar",
+            "MAG",
+            "ACL",
+            "DBLP",
+            "DOI",
+            "PubMed",
+            "PubMedCentral",
```

Comment:
This list is repeated twice in the code and we might need to update it in the future, so it should be abstracted.